src/sentry/spans/consumers/process/flusher.py

```diff
@@ -118,41 +119,50 @@ def produce(payload: KafkaPayload) -> None:
                     producer_futures.append(producer.produce(topic, payload))
 
             while not stopped.value:
-                now = int(time.time()) + current_drift.value
-                flushed_segments = buffer.flush_segments(max_segments=max_flush_segments, now=now)
+                with metrics.timer("spans.buffer.flusher.loop_body"):
+                    now = int(time.time()) + current_drift.value
+                    flushed_segments = buffer.flush_segments(
+                        max_segments=max_flush_segments, now=now
+                    )
 
-                should_backpressure.value = len(flushed_segments) >= max_flush_segments * len(
-                    buffer.assigned_shards
-                )
+                    if len(flushed_segments) >= max_flush_segments * len(buffer.assigned_shards):
+                        if backpressure_since.value == 0:
+                            backpressure_since.value = int(time.time())
+                    else:
+                        backpressure_since.value = 0
 
-                if not flushed_segments:
-                    time.sleep(1)
-                    continue
-
-                for _, flushed_segment in flushed_segments.items():
-                    if not flushed_segment.spans:
-                        # This is a bug, most likely the input topic is not
-                        # partitioned by trace_id so multiple consumers are writing
-                        # over each other. The consequence is duplicated segments,
-                        # worst-case.
-                        metrics.incr("sentry.spans.buffer.empty_segments")
+                    if not flushed_segments:
+                        time.sleep(1)
                         continue
 
-                    spans = [span.payload for span in flushed_segment.spans]
+                    with metrics.timer("spans.buffer.flusher.produce"):
```

Comment:
Isn't the call to `produce` fully async below, or can it block when the buffer is full?
Either way, a significant portion of this is the `dumps` - Can we split this timing into serializing and producing?

Reply:
it's mostly about the dumps, yeah. spawning the futures and everything that is necessary. open to suggestions to rename the metric.

`wait_produce` is then joining on those futures